networkx/drawing/nx_pylab.py

```diff
@@ -467,6 +468,180 @@ def draw_networkx_nodes(
     return node_collection
 
 
+class FancyArrowFactory:
+    """Draw arrows with `matplotlib.patches.FancyarrowPatch`"""
+
+    class ConnectionStyleFactory:
+        def __init__(self, connectionstyles, selfloop_height, ax=None):
+            import matplotlib as mpl
+            import matplotlib.path  # call as mpl.path
+            import matplotlib.pyplot as plt
+            import numpy as np
+
+            if ax is None:
+                ax = plt.gca()
+            self.ax = ax
+            self.mpl = mpl
+            self.np = np
+            self.connectionstyles = connectionstyles
+            self.base_connection_styles = [
+                mpl.patches.ConnectionStyle(cs) for cs in connectionstyles
+            ]
+            self.selfloop_height = selfloop_height
+
+        def curved(self, edge_key):
+            return self.base_connection_styles[edge_key]
+
+        def self_loop(self):
+            def connectionstyle(posA, posB, *args, **kwargs):
+                assert self.np.all(posA == posB)
```

Comment:
What happens here if the assertion doesn't hold?
This should probably be a check with a raise of `nx.NetworkXError` if not true.